File: .claude/plans/merge-convert-skills/tools/ir-extract-python/tests/fixtures/pattern_matching.py
"""Pattern matching examples (Python 3.10+) for testing."""

from dataclasses import dataclass
from typing import Any


@dataclass
class Point:
    """A 2D point for pattern matching examples."""

    x: float
    y: float


def match_literal(value: int) -> str:
    """Match against literal values."""
    match value:
        case 0:
            return "zero"
        case 1:
            return "one"
        case 2:
            return "two"
        case _:
            return "other"


def match_sequence(seq: list[int]) -> str:
    """Match against sequences."""
    match seq:
        case []:
            return "empty"
        case [x]:
            return f"single: {x}"
        case [x, y]:
            return f"pair: {x}, {y}"
        case [x, *rest]:
            return f"head: {x}, tail: {rest}"


def match_mapping(data: dict[str, Any]) -> str:
    """Match against mappings."""
    match data:
        case {"type": "error", "message": msg}:
            return f"Error: {msg}"
        case {"type": "success", "data": d}:
            return f"Success: {d}"
        case {"type": t}:
            return f"Unknown type: {t}"
        case _:
            return "Invalid data"


def match_class(obj: Any) -> str:
    """Match against class patterns."""
    match obj:
        case Point(x=0, y=0):
            return "origin"
        case Point(x=0, y=y):
            return f"on y-axis at {y}"
        case Point(x=x, y=0):
            return f"on x-axis at {x}"
        case Point(x=x, y=y):
            return f"point at ({x}, {y})"
        case _:
            return "not a point"


def match_with_guard(value: int) -> str:
    """Match with guard conditions."""
    match value:
        case n if n < 0:
            return "negative"
        case n if n == 0:
            return "zero"
        case n if n < 10:
            return "small positive"
        case n if n < 100:
            return "medium"
        case _:
            return "large"


def match_or_patterns(char: str) -> str:
    """Match with or patterns."""
    match char:
        case "a" | "e" | "i" | "o" | "u":
            return "vowel"
        case "y":
            return "sometimes vowel"
        case _:
            return "consonant"


def match_as_pattern(data: tuple[int, int]) -> str:
    """Match with as patterns."""
    match data:
        case (0, 0) as origin:
            return f"at origin: {origin}"
        case (x, y) as point if x == y:
            return f"on diagonal: {point}"
        case (x, y):
            return f"point: ({x}, {y})"


def match_nested(data: dict[str, Any]) -> str:
    """Match with nested patterns."""
    match data:
        case {"user": {"name": name, "role": "admin"}}:
            return f"Admin user: {name}"
        case {"user": {"name": name, "role": role}}:
            return f"User: {name} ({role})"
        case {"error": {"code": code, "message": msg}}:
            return f"Error {code}: {msg}"
        case _:
            return "Unknown format"


def match_complex_sequence(items: list[dict[str, Any]]) -> str:
    """Match complex sequence patterns."""
    match items:
        case []:
            return "empty"
        case [{"type": "header", "title": title}, *body, {"type": "footer"}]:
            return f"document: {title} with {len(body)} sections"
        case [{"type": "header", "title": title}, *body]:
            return f"document without footer: {title}"
        case _:
            return "malformed document"

File: .claude/plans/merge-convert-skills/tests/fixtures/scala/control/pattern_matching.scala
// Pattern Matching in Scala
//
// Demonstrates exhaustive pattern matching and extractors.

package examples.control

// Basic pattern matching
object BasicPatterns {
  def describe(x: Any): String = x match {
    case 0 => "zero"
    case 1 => "one"
    case "hello" => "greeting"
    case true => "yes"
    case false => "no"
    case _ => "something else"
  }

  // Type patterns
  def typeMatch(x: Any): String = x match {
    case i: Int => s"Int: $i"
    case s: String => s"String: $s"
    case l: List[_] => s"List with ${l.length} elements"
    case _ => "Unknown type"
  }

  // Guard patterns
  def classify(n: Int): String = n match {
    case x if x < 0 => "negative"
    case 0 => "zero"
    case x if x <= 10 => "small positive"
    case x if x <= 100 => "medium positive"
    case _ => "large positive"
  }
}

// Case class patterns
object CaseClassPatterns {
  case class Person(name: String, age: Int)
  case class Address(street: String, city: String)
  case class Employee(person: Person, address: Address)

  def greet(person: Person): String = person match {
    case Person("Alice", _) => "Hello, Alice!"
    case Person(name, age) if age < 18 => s"Hi, young $name!"
    case Person(name, _) => s"Hello, $name"
  }

  // Nested patterns
  def getCity(employee: Employee): String = employee match {
    case Employee(_, Address(_, city)) => city
  }

  // Variable binding with @
  def checkPerson(person: Person): String = person match {
    case p @ Person(_, age) if age >= 18 => s"Adult: ${p.name}"
    case p => s"Minor: ${p.name}"
  }
}

// Sealed trait patterns (exhaustive matching)
object SealedPatterns {
  sealed trait Shape
  case class Circle(radius: Double) extends Shape
  case class Rectangle(width: Double, height: Double) extends Shape
  case class Triangle(a: Double, b: Double, c: Double) extends Shape

  def area(shape: Shape): Double = shape match {
    case Circle(r) => math.Pi * r * r
    case Rectangle(w, h) => w * h
    case Triangle(a, b, c) =>
      val s = (a + b + c) / 2
      math.sqrt(s * (s - a) * (s - b) * (s - c))
  }

  // Option matching
  def process(opt: Option[Int]): String = opt match {
    case Some(n) if n > 0 => s"Positive: $n"
    case Some(n) if n < 0 => s"Negative: $n"
    case Some(0) => "Zero"
    case None => "Nothing"
  }

  // Either matching
  def handleResult(result: Either[String, Int]): String = result match {
    case Right(n) => s"Success: $n"
    case Left(err) => s"Error: $err"
  }
}

// List patterns
object ListPatterns {
  def describeList(list: List[Int]): String = list match {
    case Nil => "empty"
    case x :: Nil => s"singleton: $x"
    case x :: y :: Nil => s"pair: $x and $y"
    case x :: y :: rest => s"starts with $x, $y and ${rest.length} more"
  }

  // Pattern with list operations
  def sum(list: List[Int]): Int = list match {
    case Nil => 0
    case head :: tail => head + sum(tail)
  }

  // With guards
  def findFirst(list: List[Int], pred: Int => Boolean): Option[Int] = list match {
    case Nil => None
    case head :: _ if pred(head) => Some(head)
    case _ :: tail => findFirst(tail, pred)
  }
}

// Tuple patterns
object TuplePatterns {
  def processPair(pair: (Int, Int)): String = pair match {
    case (0, 0) => "origin"
    case (x, 0) => s"on x-axis at $x"
    case (0, y) => s"on y-axis at $y"
    case (x, y) if x == y => s"on diagonal at $x"
    case (x, y) => s"at ($x, $y)"
  }

  def swap[A, B](pair: (A, B)): (B, A) = pair match {
    case (a, b) => (b, a)
  }

  // Triple patterns
  def classify(triple: (Int, Int, Int)): String = triple match {
    case (a, b, c) if a == b && b == c => "all equal"
    case (a, b, _) if a == b => "first two equal"
    case (_, b, c) if b == c => "last two equal"
    case _ => "all different"
  }
}

// Extractor patterns
object ExtractorPatterns {
  // Custom extractor
  object Even {
    def unapply(n: Int): Option[Int] =
      if (n % 2 == 0) Some(n / 2) else None
  }

  object Odd {
    def unapply(n: Int): Boolean = n % 2 == 1
  }

  def classify(n: Int): String = n match {
    case Even(half) => s"Even, half is $half"
    case Odd() => "Odd"
  }

  // Regex extractor
  val EmailPattern = """(\w+)@(\w+)\.(\w+)""".r

  def parseEmail(email: String): Option[(String, String, String)] = email match {
    case EmailPattern(user, domain, tld) => Some((user, domain, tld))
    case _ => None
  }

  // Sequence extractor
  object Split {
    def unapplySeq(s: String): Option[Seq[String]] =
      Some(s.split(",").toSeq)
  }

  def parseCsv(line: String): List[String] = line match {
    case Split(first, second, rest @ _*) =>
      List(first, second) ++ rest
    case Split(single) => List(single)
    case Split() => Nil
  }
}

// Partial functions from patterns
object PartialFunctionPatterns {
  // Partial function literal
  val positive: PartialFunction[Int, String] = {
    case n if n > 0 => s"Positive: $n"
  }

  val negative: PartialFunction[Int, String] = {
    case n if n < 0 => s"Negative: $n"
  }

  // Combining partial functions
  val classify: PartialFunction[Int, String] =
    positive orElse negative orElse { case 0 => "Zero" }

  // Using with collect
  val numbers = List(-2, -1, 0, 1, 2)
  val positives: List[String] = numbers.collect {
    case n if n > 0 => s"Positive: $n"
  }
}

// Match expressions as values
object MatchAsValue {
  def describe(n: Int): String = {
    val kind = n match {
      case 0 => "zero"
      case x if x > 0 => "positive"
      case _ => "negative"
    }
    s"$n is $kind"
  }

  // Inline match
  val result: String = (1, 2) match {
    case (a, b) => s"sum: ${a + b}"
  }
}
